torch/_inductor/ir.py

```diff
@@ -4339,6 +4372,34 @@ def should_allocate(self):
         return True
 
 
+class OutputBuffer(ExternKernel):
+    """
+    Represent the output buffer used by ops that require multiple of them
+    """
+
+    def __init__(self, layout):
+        super().__init__(name=None, layout=layout, inputs=[])
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
+        return True
+
+    def codegen(self, wrapper):
+        wrapper.writeline(f"# collective out buffer {self.name}")
+
+
+class MultiOutputNoSizeAssert(MultiOutput):
+    """
+    Extract partial output from a multi-output OP.
+    Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
```

Comment:
i get that you want bigger sizes in the allgather case, but is not-asserting them just bc its trickier to know what to assert, or what?

Reply:
We don't need to assert because the lowering using this node emits the alloc of the exact size we're asserting here.

Since inductor doesn't have a strength reduction pass for guards that works with MultiOutput and CPU usage is a problem with Inductor, I decided to go with this.